# load the needed libraries
library(dplyr)
library(numbat)
# load the needed variables
# this is from beginning of this script
count_mat = readRDS("/broad/vangalenlab/sariipek/numbat/Pt08/pt8_host_numbat_umi_counts.rds")
# this is from the first part of Numbat from VM
pre <-read.table("/broad/vangalenlab/sariipek/numbat/Pt08/pre_allele_counts.tsv", sep = "\t", header = TRUE)
rem1 <-read.table("/broad/vangalenlab/sariipek/numbat/Pt08/rem1_allele_counts.tsv", sep = "\t", header = TRUE)
rem2 <-read.table("/broad/vangalenlab/sariipek/numbat/Pt08/rem2_allele_counts.tsv", sep = "\t", header = TRUE)
rel <-read.table("/broad/vangalenlab/sariipek/numbat/Pt08/rel_allele_counts.tsv", sep = "\t", header = TRUE)
df_allele <- rbind(pre,rem1,rem2,rel)


  # Run Numbat

  out = run_numbat(
  count_mat, # gene x cell integer UMI count matrix 
  ref_hca, # reference expression profile, a gene x cell type normalized expression level matrix
  df_allele, # allele dataframe generated by pileup_and_phase script
  genome = "hg38",
  t = 1e-5,
  min_LLR = 0.5,
  ncores = 4,
  plot = TRUE,
  out_dir = './hostpt8'
)

